subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheAccess.java

```diff
@@ -177,29 +181,34 @@ public void useCache(Runnable action) {
 
     @Override
     public <T> T useCache(Factory<? extends T> factory) {
-        boolean wasStarted;
-        lock.lock();
-        try {
-            takeOwnership();
-            wasStarted = onStartWork();
-        } finally {
-            lock.unlock();
-        }
-        try {
-            return factory.create();
-        } finally {
-            lock.lock();
+            boolean wasStarted;
+            stateLock.lock();
             try {
+                takeOwnership();
                 try {
-                    if (wasStarted) {
-                        onEndWork();
-                    }
-                } finally {
+                    wasStarted = onStartWork();
+                } catch (Throwable t) {
```

Comment:
If we don't catch `Throwable`, anything could happen so the error would surface again. We know it can happen with a timeout exception, but I wouldn't guarantee other exceptions couldn't happen too.